## Day 1

In [192]:
import pandas as pd
from scipy.io import arff

# Load the arff file
data = arff.loadarff('../data/raw/pc1.arff')

# Convert to a pandas DataFrame
df = pd.DataFrame(data[0])

print(df.head())

     loc   v(g)  ev(g)  iv(G)       N         V     L      D       I  \
0    1.1    1.4    1.4    1.4     1.3      1.30  1.30   1.30    1.30   
1    1.0    1.0    1.0    1.0     1.0      1.00  1.00   1.00    1.00   
2   91.0    9.0    3.0    2.0   318.0   2089.21  0.04  27.68   75.47   
3  109.0   21.0    5.0   18.0   381.0   2547.56  0.04  28.37   89.79   
4  505.0  106.0   41.0   82.0  2339.0  20696.93  0.01  75.93  272.58   

            E  ...  lOCode  lOComment  locCodeAndComment  lOBlank  uniq_Op  \
0        1.30  ...     2.0        2.0                2.0      2.0      1.2   
1        1.00  ...     1.0        1.0                1.0      1.0      1.0   
2    57833.24  ...    80.0       44.0               11.0     31.0     29.0   
3    72282.68  ...    97.0       41.0               12.0     24.0     28.0   
4  1571506.88  ...   457.0       71.0               48.0     49.0     64.0   

   uniq_Opnd  total_Op  total_Opnd  branchCount   defects  
0        1.2       1.2         1.2    

In [193]:
print("Shape:")
print(df.shape)

print("\nColumn names:")
print(df.columns.tolist())

print("\nData types:")
print(df.dtypes)

print("\nMissing values:")
print(df.isnull().sum())

Shape:
(1109, 22)

Column names:
['loc', 'v(g)', 'ev(g)', 'iv(G)', 'N', 'V', 'L', 'D', 'I', 'E', 'B', 'T', 'lOCode', 'lOComment', 'locCodeAndComment', 'lOBlank', 'uniq_Op', 'uniq_Opnd', 'total_Op', 'total_Opnd', 'branchCount', 'defects']

Data types:
loc                  float64
v(g)                 float64
ev(g)                float64
iv(G)                float64
N                    float64
V                    float64
L                    float64
D                    float64
I                    float64
E                    float64
B                    float64
T                    float64
lOCode               float64
lOComment            float64
locCodeAndComment    float64
lOBlank              float64
uniq_Op              float64
uniq_Opnd            float64
total_Op             float64
total_Opnd           float64
branchCount          float64
defects               object
dtype: object

Missing values:
loc                  0
v(g)                 0
ev(g)                0
iv(G)      

In [194]:
df["defects"].value_counts()

defects
b'false'    1032
b'true'       77
Name: count, dtype: int64

In [195]:
df["defects"].value_counts(normalize=True) * 100

defects
b'false'    93.056808
b'true'      6.943192
Name: proportion, dtype: float64

In [196]:
df.describe()

,loc,v(g),ev(g),iv(G),N,V,L,D,I,E,...,T,lOCode,lOComment,locCodeAndComment,lOBlank,uniq_Op,uniq_Opnd,total_Op,total_Opnd,branchCount
count,1109.000000,1109.000000,1109.000000,1109.000000,1109.000000,1109.000000,1109.000000,1109.000000,1109.000000,1.109000e+03,...,1109.000000,1109.000000,1109.000000,1109.000000,1109.000000,1109.000000,1109.000000,1109.000000,1109.000000,1109.000000
mean,23.376105,5.510730,2.766817,3.321371,117.393417,699.711217,0.129414,15.396267,32.904472,2.882288e+04,...,1601.273030,22.434626,4.696123,0.944995,6.745717,13.307665,20.892876,66.493417,50.901894,9.576555
std,35.284017,8.958976,5.567704,6.401998,197.336888,1509.545684,0.146952,16.337508,35.387685,1.706436e+05,...,9480.199989,33.574645,10.518447,3.345297,12.301210,8.182224,29.051434,111.703236,86.308689,16.540676
min,0.000000,1.000000,1.000000,1.000000,1.000000,0.000000,0.000000,0.000000,0.000000,0.000000e+00,...,0.000000,0.000000,0.000000,0.000000,0.000000,1.000000,0.000000,1.000000,0.000000,1.000000
25%,7.000000,1.000000,1.000000,1.000000,25.000000,97.670000,0.050000,5.710000,14.040000,5.581300e+02,...,31.010000,7.000000,0.000000,0.000000,1.000000,8.000000,6.000000,15.000000,10.000000,1.000000
50%,13.000000,3.000000,1.000000,2.000000,58.000000,275.100000,0.080000,11.610000,23.570000,3.189170e+03,...,177.180000,13.000000,0.000000,0.000000,2.000000,12.000000,12.000000,33.000000,24.000000,5.000000
75%,26.000000,6.000000,3.000000,3.000000,126.000000,674.040000,0.160000,20.500000,41.150000,1.238156e+04,...,687.860000,24.000000,5.000000,1.000000,8.000000,17.000000,25.000000,72.000000,56.000000,11.000000
max,602.000000,136.000000,123.000000,123.000000,2785.000000,25942.690000,2.000000,270.660000,598.330000,4.279633e+06,...,237757.400000,600.000000,159.000000,48.000000,225.000000,99.000000,538.000000,1641.000000,1144.000000,236.000000


## Initial EDA Findings
* The class distribution of the data set is imbalanced.
* The PC1 dataset consists of 1109 rows and 22 columns. Rows contain information on software modules and columns contain attributes of teh respective software modules.
* There are 22 attribute columns out of which 21 columns are used as explanatory variates (input of the model) and 1 is response variate (output of the model).
* There are no missing values found in the dataset.
* There are total of 1032 non-defective modules and 77 defective modules out of 1109 software modules.
* Several explanatory variables are rightly skewed because their means are larger than their medians.

## Day 2

In [197]:
X = df.drop(columns=["defects"])
Y = df["defects"]

print(X.shape)
print(Y.shape)

(1109, 21)
(1109,)


In [198]:
from sklearn.model_selection import train_test_split

In [199]:
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.2, random_state=42, stratify=Y)

In [200]:
print(X_train.shape)
print(X_test.shape)
print(Y_train.shape)
print(Y_test.shape)

(887, 21)
(222, 21)
(887,)
(222,)


In [201]:
Y.value_counts(normalize=True)
Y_train.value_counts(normalize=True)
Y_test.value_counts(normalize=True)

defects
b'false'    0.932432
b'true'     0.067568
Name: proportion, dtype: float64

## Day 3

In [202]:
from sklearn.dummy import DummyClassifier
from sklearn.metrics import accuracy_score, confusion_matrix, precision_score, recall_score, f1_score, roc_auc_score
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.pipeline import make_pipeline

In [203]:
Y_train = Y_train.astype(str)
Y_test = Y_test.astype(str)

dummy = DummyClassifier(strategy="most_frequent")

In [204]:
dummy.fit(X_train, Y_train)
dummy_prediction = dummy.predict(X_test)

print(dummy_prediction[:20])

['false' 'false' 'false' 'false' 'false' 'false' 'false' 'false' 'false'
 'false' 'false' 'false' 'false' 'false' 'false' 'false' 'false' 'false'
 'false' 'false']


In [205]:
dummyAccuracy = accuracy_score(Y_test, dummy_prediction)
dummyConfusionMatrix = confusion_matrix(Y_test, dummy_prediction)

print("Accuracy:", dummyAccuracy)
print("Confusion Matrix:")
print(dummyConfusionMatrix)

Accuracy: 0.9324324324324325
Confusion Matrix:
[[207   0]
 [ 15   0]]


In [206]:
precision = precision_score(Y_test, dummy_prediction, pos_label="true", zero_division=0)
recall = recall_score(Y_test, dummy_prediction, pos_label="true")
f1 = f1_score(Y_test, dummy_prediction, pos_label="true", zero_division=0)

print("Precision:", precision)
print("Recall:", recall)
print("F1-score:", f1)

Precision: 0.0
Recall: 0.0
F1-score: 0.0


In [207]:
pipe = make_pipeline(StandardScaler(), LogisticRegression())

In [208]:
pipe.fit(X_train, Y_train)
pipe_prediction = pipe.predict(X_test)
pipeProbability = pipe.predict_proba(X_test)
pipe_defectiveProbability = pipeProbability[:, 1]

print(pipe_prediction[:20])

print("\nClasses:")
print(pipe.classes_)

print("\nProbability shape:")
print(pipeProbability.shape)

print("\nFirst 5 probability rows:")
print(pipeProbability[:5])

['false' 'false' 'false' 'false' 'false' 'false' 'false' 'false' 'false'
 'false' 'false' 'false' 'false' 'false' 'false' 'false' 'false' 'false'
 'false' 'false']

Classes:
['false' 'true']

Probability shape:
(222, 2)

First 5 probability rows:
[[0.97725754 0.02274246]
 [0.97589319 0.02410681]
 [0.97895044 0.02104956]
 [0.9662071  0.0337929 ]
 [0.94999157 0.05000843]]


In [209]:
pipeAccuracyScore = accuracy_score(Y_test, pipe_prediction)
pipeConfusionMatrix = confusion_matrix(Y_test, pipe_prediction)
pipePrecision = precision_score(Y_test, pipe_prediction, pos_label="true", zero_division=0)
pipeRecall = recall_score(Y_test, pipe_prediction, pos_label="true")
pipeF1 = f1_score(Y_test, pipe_prediction, pos_label="true", zero_division=0)
pipeROC_AUC = roc_auc_score(Y_test, pipe_defectiveProbability)

print("Accuracy: ", pipeAccuracyScore)
print("\nConfusion Matrix: ")
print(pipeConfusionMatrix)

print("\nPrecision:", pipePrecision)
print("\nRecall:", pipeRecall)
print("\nF1-score:", pipeF1)
print("\nROC-AUC Score:", pipeROC_AUC)

Accuracy:  0.9279279279279279

Confusion Matrix: 
[[205   2]
 [ 14   1]]

Precision: 0.3333333333333333

Recall: 0.06666666666666667

F1-score: 0.1111111111111111

ROC-AUC Score: 0.864573268921095


## Day 3 Findings

The DummyClassifier baseline achieved about 93.2% accuracy but detected 0 defective modules. Logistic Regression achieved about 92.8% accuracy and detected 1 defective modules, but it still missed 14 of the 15 defective modules in the test set. This shows that Logistic Regression does not outperform the baseline in terms of accuracy, but it does improve defect detection by identifying one defective module that the baseline missed. Therefore, accuracy alone does not give a complete picture of the model's performance, especially because the dataset is imbalanced.

## Day 4

In [210]:
from sklearn.ensemble import RandomForestClassifier

In [211]:
randomForest = RandomForestClassifier(random_state=42)

In [212]:
randomForest.fit(X_train, Y_train)

randomForest_prediction = randomForest.predict(X_test)
randomForest_probability = randomForest.predict_proba(X_test)
randomForest_defectiveProbability =randomForest_probability[:, 1]

print(randomForest_prediction[:20])

print("\nClasses:")
print(randomForest.classes_)

print("\nProbability shape:")
print(randomForest_probability.shape)

print("\nFirst 5 probability rows:")
print(randomForest_probability[:5])


['false' 'false' 'false' 'false' 'false' 'false' 'false' 'false' 'false'
 'false' 'false' 'false' 'false' 'false' 'false' 'true' 'false' 'false'
 'false' 'false']

Classes:
['false' 'true']

Probability shape:
(222, 2)

First 5 probability rows:
[[1. 0.]
 [1. 0.]
 [1. 0.]
 [1. 0.]
 [1. 0.]]


In [213]:
randomForest_confusionMatrix = confusion_matrix(Y_test, randomForest_prediction)
randomForest_accuracy = accuracy_score(Y_test, randomForest_prediction)
randomForest_precision = precision_score(Y_test, randomForest_prediction, pos_label="true", zero_division=0)
randomForest_recall = recall_score(Y_test, randomForest_prediction, pos_label="true", zero_division=0)
randomForest_F1 = f1_score(Y_test, randomForest_prediction, pos_label="true", zero_division=0)
randomForest_ROC_AUC = roc_auc_score(Y_test, randomForest_defectiveProbability)

print("Confusion Matrix: ")
print(randomForest_confusionMatrix)
print("\nAccuracy: ", randomForest_accuracy)
print("\nPrecision:", randomForest_precision)
print("\nRecall:", randomForest_recall)
print("\nF1-score:", randomForest_F1)
print("\nROC-AUC Score:", randomForest_ROC_AUC)

Confusion Matrix: 
[[204   3]
 [ 10   5]]

Accuracy:  0.9414414414414415

Precision: 0.625

Recall: 0.3333333333333333

F1-score: 0.43478260869565216

ROC-AUC Score: 0.8750402576489533


## Day 4 Findings

Random Forest achieved higher accuracy (94.1%) than Logistic Regression (92.8%). Random Forest detected 5 of the 15 defective modules, while Logistic Regression detected only 1 of 15. This means Random Forest achieved a higher recall of 33.3%, compared with 6.7% for Logistic Regression.

Random Forest had higher precision, F1-score, and ROC-AUC than Logistic Regression. Therefore, Random Forest performed better than Logistic Regression on PC1 under this train/test split.

However, Random Forest still missed 10 of the 15 defective modules. This means that although Random Forest performed better than Logistic Regression, it still did not detect a large proportion of the actual defects. This shows accuracy alone is not sufficient for evaluating software defect prediction models.